packages/core/saved-objects/core-saved-objects-migration-server-internal/src/zdt/utils/generate_additive_mapping_diff.test.ts

```diff
@@ -136,11 +154,85 @@ describe('generateAdditiveMappingDiff', () => {
     expect(() =>
       generateAdditiveMappingDiff({
         types,
-        meta,
+        mapping: mappingFromMeta(meta),
         deletedTypes,
       })
     ).toThrowErrorMatchingInlineSnapshot(
       `"Cannot generate additive mapping diff: mappingVersions not present on index meta"`
     );
   });
+
+  it('throws an error if _meta is not present on the index', () => {
+    expect(() =>
+      generateAdditiveMappingDiff({
+        types: [],
+        mapping: {
+          properties: {},
+        },
+        deletedTypes: [],
+      })
+    ).toThrowErrorMatchingInlineSnapshot(
+      `"Cannot generate additive mapping diff: meta not present on index"`
+    );
+  });
+
+  it('includes the root fields that were modified', () => {
+    const { foo, bar } = getTypes();
+    const types = [foo, bar];
+    const meta: IndexMappingMeta = {
+      mappingVersions: {
+        foo: '10.2.0',
+        bar: '8.5.0',
+      },
+    };
+
+    getBaseMappingsMock.mockReturnValue({
+      properties: {
+        rootA: { type: 'keyword' },
+        rootB: { type: 'keyword' },
+      },
+    });
+    getUpdatedRootFieldsMock.mockReturnValue(['rootA']);
+
+    const addedMappings = generateAdditiveMappingDiff({
+      types,
+      mapping: mappingFromMeta(meta),
+      deletedTypes,
+    });
+
+    expect(addedMappings).toEqual({
+      rootA: { type: 'keyword' },
+    });
+  });
+
+  it('combines the changes from the types and from the root fields', () => {
```

Comment:
Could we test behavior when adding properties to existing root fields, e.g. `references`?

Reply:
Not sure exactly how to do given helpers are stubbed for those unit tests, but I added one in https://github.com/elastic/kibana/pull/179595/commits/54d89d87ea5c929acfa0185d213581d94cc8bb60, is that what you had in mind?

TBH this additive mapping diff generation is kinda unnecessary, we could just push everything as long as we know at least one change was done, but now that we have it...

Reply:
Yes, that's what I had in mind. The test covers it, thanks!